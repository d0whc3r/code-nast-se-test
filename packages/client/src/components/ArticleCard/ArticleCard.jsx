import './ArticleCard.styles.css'

function ArticleCard({ article }) {
  const showDateTime = (dateString) => {
    const date = new Date(dateString)
    return date.toLocaleString()
  }

  return (
    <a href={article.url} target="_blank" className="card">
      <header>
        <h2 className="title">{article.title}</h2>
      </header>
      <img className="image" src={article.urlToImage} alt={article.title} />
      <p>{article.content}</p>
      <p className="published">{showDateTime(article.publishedAt)}</p>
    </a>
  )
}

export default ArticleCard
