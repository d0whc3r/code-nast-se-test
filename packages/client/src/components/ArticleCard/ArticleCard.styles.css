.card {
    @apply relative block px-6 py-3 pb-10 max-w-sm bg-white rounded-lg border border-gray-200 shadow-md hover:bg-gray-100;
}

.title {
    @apply mb-4 text-2xl font-bold tracking-tight text-gray-900 break-normal;
}

.image {
    @apply mb-4 rounded-lg;
}

.published {
    @apply absolute bottom-3 right-6 text-xs text-gray-600 text-right;
}
