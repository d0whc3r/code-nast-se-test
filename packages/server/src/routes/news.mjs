import fetch from 'node-fetch'
import { generateUrl } from '../common.mjs'

const PAGE_SIZE = 10

/**
 * Wrapper function to get news and return in http response
 * @param params {Record<string, string>}
 * @param res {import('http').ServerResponse}
 */
async function getNews(params, res) {
  const page = +params.page || 1
  const pageSize = +params.pageSize || PAGE_SIZE
  const url = generateUrl({ ...params, pageSize: PAGE_SIZE })
  return fetch(url)
    .then((response) => response.json())
    .then((data) => {
      if (data.status === 'ok') {
        res.status(200).json({ ...data, pageSize, page, totalPages: Math.ceil(data.totalResults / PAGE_SIZE) })
      } else {
        res.status(500).json({ error: data.message })
      }
    })
    .catch((error) => {
      res.status(500).json({ error })
    })
}

/**
 * Get a list of news by topic
 * @param req {import('http').IncomingMessage}
 * @param res {import('http').ServerResponse}
 * @param next {import('express').NextFunction}
 */
export async function newsByTopic(req, res, next) {
  const params = { ...req.params, ...req.query }
  return getNews(params, res)
}

/**
 * Get a list of news by topic
 * @param req {import('http').IncomingMessage}
 * @param res {import('http').ServerResponse}
 * @param next {import('express').NextFunction}
 */
export async function latestNews(req, res, next) {
  const today = new Date().toISOString().split('T')[0]
  const yesterday = new Date(new Date().setDate(new Date().getDate() - 1)).toISOString().split('T')[0]
  const params = { sources: 'engadget', from: yesterday, to: today, ...req.query }
  return getNews(params, res)
}
