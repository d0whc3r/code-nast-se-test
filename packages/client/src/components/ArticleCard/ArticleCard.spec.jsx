import { describe, expect, it } from 'vitest'
import { render } from '../../../test/utils'
import { mockedNews } from '../../api/news.mock'
import ArticleCard from './ArticleCard'

const article = mockedNews.articles[0]

describe('ArticleCard', () => {
  it('should render', () => {
    const { baseElement } = render(<ArticleCard article={article} />)
    expect(baseElement).toMatchSnapshot()
  })
})
