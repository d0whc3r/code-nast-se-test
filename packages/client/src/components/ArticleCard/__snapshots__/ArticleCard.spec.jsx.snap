// Vitest Snapshot v1

exports[`ArticleCard > should render 1`] = `
<body>
  <div>
    <a
      class="card"
      href="https://www.enriquedans.com/2022/05/rusia-y-las-vpn-como-sintoma.html"
      target="_blank"
    >
      <header>
        <h2
          class="title"
        >
          Rusia y las VPN como síntoma
        </h2>
      </header>
      <img
        alt="Rusia y las VPN como síntoma"
        class="image"
        src="https://www.enriquedans.com/wp-content/uploads/2022/05/VPN.jpeg"
      />
      <p>
        Las descargas y el uso de redes privadas virtuales (VPNs) en Rusia se han disparado desde el inicio de la brutal invasión de Ucrania, con niveles que se aproximan al medio millón de descargas diario … [+3209 chars]
      </p>
      <p
        class="published"
      >
        07/05/2022, 16:46:50
      </p>
    </a>
  </div>
</body>
`;
